proxy.go

```
@@ -55,39 +51,79 @@ func (h *proxyHandler) handleConnect(proxyResp http.ResponseWriter, proxyReq *ht
 		sendError(proxyResp, "Failed to hijack connection", http.StatusInternalServerError, err)
 		return
 	}
-
-	log.Debug("hijacked connection, writing connect request")
-
 	// We're responsible for closing the client connection since we hijacked it. But since we're
 	// just piping data back and forth, we can give that responsibility back to the server, which
 	// will close the connection when the request is done.
 	context.AfterFunc(proxyReq.Context(), func() { clientConn.Close() })
+	log.Debug("hijacked connection")
+
+	if h.proxylessDialer != nil {
+		err := h.tryProxylessConnect(proxyReq, clientConn)
+		if err == nil {
+			return
+		}
+		log.Debugf("failed with proxyless connection: %w", err)
+	}
+
+	targetConn, err := h.dialer.DialStream(proxyReq.Context(), h.addr)
+	if err != nil {
+		sendError(proxyResp, "Failed to dial target", http.StatusServiceUnavailable, err)
+		return
+	}
+	defer targetConn.Close()
 
 	// Create a new CONNECT request to send to the proxy server.
 	connectReq, err := backend.NewRequestWithHeaders(
 		proxyReq.Context(),
 		http.MethodConnect,
 		proxyReq.URL.String(),
-		nil,
+		http.NoBody,
 	)
+	if err != nil {
+		sendError(proxyResp, "Error creating connect request", http.StatusInternalServerError, err)
+		return
+	}
 	connectReq.Header.Set(authTokenHeader, h.authToken)
+
 	if err = connectReq.Write(targetConn); err != nil {
 		sendError(proxyResp, "Failed to write connect request to proxy", http.StatusInternalServerError, err)
 		return
 	}
+	h.pipeData(targetConn, clientConn)
```

This error is ignored -- I think it should be logged if it's not an EOF.